packages/cli/src/cmds/preview/preview.ts

```diff
@@ -289,6 +309,171 @@ export default class Preview extends YargsCommand {
     return ok(null);
   }
 
+  private async spfxPreviewSetup(spfxRoot: string): Promise<Result<null, FxError>> {
+    // init service log writer
+    this.serviceLogWriter = new ServiceLogWriter();
+    await this.serviceLogWriter.init();
+
+    // run npm install for spfx
+    const spfxInstallTask = new Task(
+      constants.spfxInstallTitle,
+      false,
+      constants.npmInstallCommand,
+      undefined,
+      {
+        shell: true,
+        cwd: spfxRoot,
+      }
+    );
+
+    const spfxInstallBar = CLIUIInstance.createProgressBar(constants.spfxInstallTitle, 1);
+    const spfxInstallStartCb = commonUtils.createTaskStartCb(
+      spfxInstallBar,
+      constants.spfxInstallStartMessage,
+      this.telemetryProperties
+    );
+    const spfxInstallStopCb = commonUtils.createTaskStopCb(
+      spfxInstallBar,
+      constants.spfxInstallSuccessMessage,
+      this.telemetryProperties
+    );
+
+    let result = await spfxInstallTask?.wait(spfxInstallStartCb, spfxInstallStopCb);
+    if (result.isErr()) {
+      return err(result.error);
+    }
+
+    // run gulp trust-dev-cert
+    const gulpCertTask = new Task(
+      constants.gulpCertTitle,
+      false,
+      constants.nodeCommand,
+      [`${spfxRoot}/node_modules/gulp/bin/gulp.js`, "trust-dev-cert", "--no-color"],
+      {
+        shell: false,
+        cwd: spfxRoot,
+      }
+    );
+
+    const gulpCertBar = CLIUIInstance.createProgressBar(constants.gulpCertTitle, 1);
+    const gulpCertStartCb = commonUtils.createTaskStartCb(
+      gulpCertBar,
+      constants.gulpCertStartMessage,
+      this.telemetryProperties
+    );
+    const gulpCertStopCb = commonUtils.createTaskStopCb(
+      gulpCertBar,
+      constants.gulpCertSuccessMessage,
+      this.telemetryProperties
+    );
+
+    result = await gulpCertTask?.wait(gulpCertStartCb, gulpCertStopCb);
+    if (result.isErr()) {
+      return err(result.error);
+    }
+
+    // run gulp serve
+    const gulpServeTask = new Task(
+      constants.gulpServeTitle,
+      true,
+      constants.nodeCommand,
+      [`${spfxRoot}/node_modules/gulp/bin/gulp.js`, "serve", "--nobrowser", "--no-color"],
+      {
+        shell: false,
+        cwd: spfxRoot,
+      }
+    );
+
```

Comment:
Add `this.backgroundTasks.push(gulpServeTask);` here.

Reply:
resolved. thx.